### ISSUE TITLE
Improve handling of rubbish in ghcWithPackages

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#30304]
## Issue description

When you pass something bogus into `ghcWithPackages`, it succeeds instead of failing loudly.

### Steps to reproduce

Write

```
nix-build --expr 'with import ./. {}; with import ./pkgs/development/haskell-modules/lib.nix { inherit pkgs lib; }; haskellPackages.ghcWithPackages (pkgs: with pkgs; [ dontCheck consul-haskell ])'
```

Debug an hour why it doesn't work (it builds the package with tests, ignoring `dontCheck`).

Realise that `[ dontCheck consul-haskell ]` should have been `[ (dontCheck consul-haskell) ]`.

---

From @cleverca22 

> i would also expect ghcWithPackages to fail hard, because you gave it a lambda when it was expecting a string/package
> so i'm surprised it even tried to build it

```
[clever@amd-nixos:~/apps/nixpkgs]$ nix-repl .
Welcome to Nix version 1.11.14. Type :? for help.

Loading '.'...
Added 7948 variables.

nix-repl> builtins.unsafeGetAttrPos "dontCheck" haskell.lib
{ column = 3; file = "/home/clever/apps/nixpkgs/pkgs/development/haskell-modules/lib.nix"; line = 24; }
```

```
nix-repl> builtins.unsafeGetAttrPos "ghcWithPackages" haskellPackages
{ column = 5; file = "/home/clever/apps/nixpkgs/pkgs/development/haskell-modules/make-package-set.nix"; line = 164; }
```

> pkgs/development/haskell-modules/with-packages-wrapper.nix

`paths         = lib.filter (x: x ? isHaskellLibrary) (lib.closePropagation packages);`

> aha

```
nix-repl> haskellPackages.consul-haskell.isHaskellLibrary
true
```

> if you give `ghcWithPackages` anything that lacks a `.isHaskellLibrary` attribute, it will silently filter it out of the package list

---

## Technical details

* Nixpkgs version: dea58ce228e6f9686773745024eb844d8dc39fc3

**Comment by nh2** [comment]
CC @cleverca22 @peti -- Ideally we could make it fail loudly on such errors.

**Comment by peti** [comment]
Yeah, that's a good point. I suppose we could throw an error if any non-Haskell packages are included in the package list.

**Comment by domenkozar** [comment]
Filtering non-Haskell packages is a bit too agressive, one might have developer tools installed around.

I suggest we assert on non-derivations to catch errors like this.

**Comment by qweered** [comment]
Opened #515252 with fix, would appreciate review and merge
